jbpm-task-assigning/jbpm-task-assigning-core/src/test/java/org/jbpm/task/assigning/model/solver/realtime/BaseProblemFactChangeTest.java

```diff
@@ -0,0 +1,149 @@
+        List<ProgrammedProblemFactChange> scheduledChanges = new ArrayList<>();
+
+        int totalProgrammedChanges = programmedChanges.size();
+        int[] pendingChanges = {programmedChanges.size()};
+
+        solver.addEventListener(event -> {
+            if (initialSolution[0] == null) {
+                //store the first produced solution for knowing how things looked like at the very beginning.
+                initialSolution[0] = event.getNewBestSolution();
+                //let the problem fact changes start being produced.
+                programNextChange.release();
+            } else if (changesInProgress[0] && event.isEveryProblemFactChangeProcessed()) {
+                changesInProgress[0] = false;
+                ProgrammedProblemFactChange programmedChange = scheduledChanges.get(scheduledChanges.size() - 1);
+                programmedChange.setSolutionAfterChange(event.getNewBestSolution());
+
+                if (pendingChanges[0] > 0) {
+                    //let the Programmed changes producer produce next change
+                    programNextChange.release();
+                } else {
+                    solver.terminateEarly();
+                    allChangesWereProduced.release();
+                }
+            }
+        });
+
+        //Programmed changes producer Thread.
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        executor.execute(() -> {
+            boolean hasMoreChanges = true;
+            while (hasMoreChanges) {
+                try {
+                    //wait until next problem fact change can be added to the solver.
+                    //by construction the lock is only released when no problem fact change is in progress.
+                    executorThreadStarted.release();
+                    programNextChange.acquire();
+                    ProgrammedProblemFactChange programmedChange = programmedChanges.remove(0);
+                    hasMoreChanges = !programmedChanges.isEmpty();
+                    pendingChanges[0] = programmedChanges.size();
+                    scheduledChanges.add(programmedChange);
+                    changesInProgress[0] = true;
+                    solver.addProblemFactChange(programmedChange.getChange());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+            try {
+                //wait until the solver listener has processed all the changes.
+                allChangesWereProduced.acquire();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        });
+
+        executorThreadStarted.acquire();
+        solver.solve(solution);
+
+        assertTrue(programmedChanges.isEmpty());
+        assertEquals(totalProgrammedChanges, scheduledChanges.size());
+        assertEquals(0, pendingChanges[0]);
+        return initialSolution[0];
```

Comment:
I want to keep the initial solution because in scenarios (runs) when I print the produced solutions during the test execution in files, I can see "how the solution looked like" before applying any of the changes. And then I print how the solution looks like after applying change#1, change#2 and so on. I was very useful to me during development to have a visual tracking that things worked well. By default the files printing is disabled so it's no a problem to keep this.